### PR TITLE
Replace Twig with v2

### PR DESCRIPTION
**Pull request by martin-georgiev**
* Drop PHP 7.0 support
* Support Twig 2

### CODE REVIEW

File: src/Entity/Template.php [hunk 1]
```diff
@@ -90,7 +90,7 @@ public function getLastModified(): \DateTimeInterface
         return $this->lastModified;
     }
 
-    public function setLastModifiedToCurrentMoment()
+    public function setLastModifiedToCurrentMoment(): void
     {
         $this->lastModified = new \DateTimeImmutable();
```
Comment (ben-challis) [inline src/Entity/Template.php]: Not a huge fan of the fact that Doctrine will load this as a `\DateTime` and we're setting it as a `\DateTimeImmutable` in places here, can we consider changing the signatures to either of the concrete classes and using one only?

Not a fan of the interface as a return value as means all consumers are unaware of it being mutable/immutable which is limiting and usually end up just converting it to a new instance to have sane side-effect free control of it. Accepting one as an argument is OK as you have control over the conversion and simplifies usage of the API.

File: src/Helper/DatabaseHelper.php [hunk 2]
```diff
@@ -19,14 +19,14 @@ public function __construct(Connection $connection, string $migrationsDir)
         $this->migrationsDir = $migrationsDir;
     }
 
-    public function cleanDatabase()
+    public function cleanDatabase(): void
```
Comment (ben-challis) [inline src/Helper/DatabaseHelper.php]: Can we move this file to the tests namespace?

File: src/Helper/DatabaseHelper.php [hunk 3]
```diff
@@ -35,15 +35,15 @@ private function createDatabase()
...
     }
 
-    private function dropDatabase()
+    private function dropDatabase(): void
     {
         $this->connection->getSchemaManager()->dropDatabase($this->connection->getDatabase());
     }
 
     /**
      * Runs the command doctrine:migrations:migrate
```
Comment (ben-challis) [inline src/Helper/DatabaseHelper.php]: Can we remove this? Isn't entirely accurate.

File: src/Loader/DatabaseTwigLoader.php [hunk 4]
```diff
@@ -18,50 +18,82 @@ public function __construct(EntityManagerInterface $entityManager, string $entit
...
     {
-        $source = $this->getValue('source', $name);
-        if (!is_string($source) || mb_strlen($source) < 1) {
-            throw new \Twig_Error_Loader(sprintf('Template "%s" does not exist.', $name));
+        return $name;
+    }
+
+    public function exists($templateName): bool
+    {
+        $template = $this->findTemplate($templateName);
```
Comment (ben-challis) [inline src/Loader/DatabaseTwigLoader.php]: Probably worth changing this as a query that checks existence as quick to do

File: src/Loader/DatabaseTwigLoader.php [hunk 5]
```diff
@@ -18,50 +18,82 @@ public function __construct(EntityManagerInterface $entityManager, string $entit
...
+                ->getQuery()
+                ->setParameter('name', $templateName)
+                ->getSingleResult();
+        } catch (NoResultException $e) {
+            return null;
         }
+    }
 
-        return $source;
+    public function getSourceContext($templateName): \Twig_Source
```
Comment (ben-challis) [inline src/Loader/DatabaseTwigLoader.php]: Can we ensure we're using the new namespaced versions for these classes for easier future compatibility? This goes across most of the bundle.

File: src/Loader/DatabaseTwigLoader.php [hunk 6]
```diff
@@ -18,50 +18,82 @@ public function __construct(EntityManagerInterface $entityManager, string $entit
         $this->entity = $entity;
     }
 
-    public function getSource($name): string
+    public function getCacheKey($name): string
     {
-        $source = $this->getValue('source', $name);
-        if (!is_string($source) || mb_strlen($source) < 1) {
-            throw new \Twig_Error_Loader(sprintf('Template "%s" does not exist.', $name));
+        return $name;
```
Comment (ben-challis) [inline src/Loader/DatabaseTwigLoader.php]: interface doesn't enforce $name to be a string so we'll need to string cast here else this can error out.

File: src/Loader/DatabaseTwigLoader.php [hunk 7]
```diff
@@ -18,50 +18,82 @@ public function __construct(EntityManagerInterface $entityManager, string $entit
...
+
+        return new \Twig_Source($templateSource, $templateName);
     }
 
-    public function getCacheKey($name): string
+    public function getSource($templateName): string
     {
-        return $name;
+        $source = $this->getValue('source', $templateName);
+        if (!is_string($source) || mb_strlen($source) < 1) {
```
Comment (ben-challis) [inline src/Loader/DatabaseTwigLoader.php]: First condition should never be true, surely? Fundamental entity mapping issue. Not convinced that an empty template should be reported as a template does not exist, an empty twig template is valid.

File: src/Loader/DatabaseTwigLoader.php [hunk 8]
```diff
@@ -18,50 +18,82 @@ public function __construct(EntityManagerInterface $entityManager, string $entit
...
+     * @return null|mixed
      */
-    private function getValue(string $column, string $templateName)
+    private function getValue(string $coloumn, string $templateName)
     {
-        $value = null;
-
         try {
-            $value = $this->entityManager
+            return $this->entityManager
```
Comment (ben-challis) [inline src/Loader/DatabaseTwigLoader.php]: Thoughts on just replacing this entirely with `findTemplate($name)->getProperty()` throughout (2 usages)? Simplifies things, can type hint properly and one of the usages is `isFresh` which is debug mode and doubt performance is much.

File: src/DependencyInjection/Compiler/TwigLoaderPass.php [hunk 9]
```diff
@@ -35,7 +35,7 @@ public function process(ContainerBuilder $container)
         $twigDefinition->replaceArgument(0, $twigChainLoaderDefinition);
     }
 
-    private function validateServices(ContainerBuilder $container, array $services)
+    private function validateServices(ContainerBuilder $container, array $services): void
```
Comment (ben-challis) [inline src/DependencyInjection/Compiler/TwigLoaderPass.php]: Relates to lives above:

I've always felt this bundle resetting your loader chain to be a bit disruptive, we could instead retrieve the current arguments and just append the database loader onto the end of it instead?

File: composer.json [hunk 10]
```diff
@@ -41,36 +41,45 @@
     "symfony/framework-bundle": "^2.7|^3.0|^4.0",
     "symfony/symfony": "^2.7|^3.0|^4.0",
     "symfony/twig-bundle": "^2.7|^3.0|^4.0",
-    "twig/twig": "^1.26"
+    "twig/twig": "^2.0"
```
[the remark this answers is not in the document]
Comment (martin-georgiev) [reply composer.json]: Good call

File: phpunit.xml [hunk 11]
```diff
@@ -1,24 +1,23 @@
...
-         colors="true"
-         bootstrap="vendor/autoload.php"
->
-    <php>
-        <ini name="error_reporting" value="-1" />
-    </php>
+<phpunit
+        colors="true"
+        convertErrorsToExceptions="true"
+        convertNoticesToExceptions="false"
```
[the remark this answers is not in the document]
Comment (martin-georgiev) [reply phpunit.xml]: done

File: src/Entity/Template.php [hunk 12]
```diff
@@ -22,7 +22,7 @@ class Template
     protected $source;
 
     /**
-     * @var array
+     * @var string[]|null
      */
     protected $services;
```
[the remark this answers is not in the document]
Comment (martin-georgiev) [reply src/Entity/Template.php]: I believe this can be dropped entirely, doesn't look like it was in good use at any point.

File: src/Loader/DatabaseTwigLoader.php (hunk 6 above)
Comment (martin-georgiev) [reply src/Loader/DatabaseTwigLoader.php]: maybe it is a good idea to see this erroring, `Twig_LoaderInterface` has old-school soft enforcement via phpDoc block which is not sufficient in a strict type world

lmk your thoughts

File: src/Loader/DatabaseTwigLoader.php (hunk 7 above)
Comment (martin-georgiev) [reply src/Loader/DatabaseTwigLoader.php]: This is legacy from Twig v1 times. One's custom mapping may declare it as nullable (e.g. our private legacy implementation). As this couldn't be enforced on method signature level (due to the old Twig v1 interface) change above takes care that the source value is useful. I've killed the method and moved the check to `getSourceContext()`.

File: src/Loader/DatabaseTwigLoader.php (hunk 8 above)
Comment (martin-georgiev) [reply src/Loader/DatabaseTwigLoader.php]: This will limit the bundle in a sense that people won't be able anymore to fully overwrite the entity. I think it is more fragile to presume a getter method with a very specific name.

Comment (ben-challis) [reply src/Loader/DatabaseTwigLoader.php]: it's already limiting by enforcing properties with exact names, we could introduce an interface and work vs that to allow bridging expected getters onto custom implementations. Let's think about this as a separate piece of work, fine as is for now.

File: src/Helper/DatabaseHelper.php (hunk 2 above)
Comment (martin-georgiev) [reply src/Helper/DatabaseHelper.php]: I think we can drop it entirely

File: src/Entity/Template.php (hunk 1 above)
Comment (martin-georgiev) [reply src/Entity/Template.php]: agree, this is updated now

File: .travis.yml [hunk 13]
```diff
@@ -7,38 +7,35 @@ cache:
...
 php:
-  - 7.0
   - 7.1.18
   - 7.2
   - nightly
 
 env:
   global:
-    - MAINTAINED_SYMFONY_VERSIONS="2.8.*|3.4.*|4.0.*|4.1.*"
+    - MAINTAINED_SYMFONY_VERSIONS="2.8.*|3.4.*|4.1.*"
```
Comment (kunicmarko20) [inline .travis.yml]: should be 4.2

Comment (kunicmarko20) [reply .travis.yml]: also, 2.8 is dead.

File: README.md [hunk 14]
```diff
@@ -1,12 +1,16 @@
-[![Build Status](https://api.travis-ci.org/Lendable/TwigBundle.svg)](https://travis-ci.org/Lendable/TwigBundle)
+[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/Lendable/TwigBundle/badges/quality-score.png)](https://scrutinizer-ci.com/g/Lendable/TwigBundle/?branch=master)
+[![Build Status](https://api.travis-ci.org/Lendable/TwigBundle.svg?branch=master)](https://www.travis-ci.org/Lendable/TwigBundle)
+[![Coverage Status](https://coveralls.io/repos/github/Lendable/TwigBundle/badge.svg?branch=master)](https://coveralls.io/github/Lendable/TwigBundle?branch=master)
+[![Latest Stable Version](https://poser.pugx.org/lendable/twig-bundle/version)](https://packagist.org/packages/lendable/twig-bundle)
+[![Total Downloads](https://poser.pugx.org/lendable/twig-bundle/downloads)](https://packagist.org/packages/lendable/twig-bundle)
 
 TwigBundle
-==========
+===
```
Comment (kunicmarko20) [inline README.md]: why this change?

Comment (kunicmarko20) [reply README.md]: same question for rest of them

File: composer.json (hunk 10 above)
Comment (kunicmarko20) [inline composer.json]: PR says "support twig 2.0" but you are dropping "1.0" should be something like "replace old twig version"

File: src/Loader/DatabaseTwigLoader.php [hunk 15]
```diff
@@ -18,50 +18,75 @@ public function __construct(EntityManagerInterface $entityManager, string $entit
...
+        }
     }
 
-    public function isFresh($name, $time): bool
+    public function getSourceContext($templateName): \Twig_Source
     {
-        if (false === $lastModified = $this->getValue('lastModified', $name)) {
+        $templateSource = $this->getValue('source', $templateName);
+        if (!is_string($templateSource) || mb_strlen($templateSource) < 1) {
+            throw new \Twig_Error_Loader(sprintf('Template "%s" does not exist.', $templateName));
```
Comment (kunicmarko20) [inline src/Loader/DatabaseTwigLoader.php]: Use Twig 2.x namespace?

File: src/Loader/DatabaseTwigLoader.php [hunk 16]
```diff
@@ -18,50 +18,75 @@ public function __construct(EntityManagerInterface $entityManager, string $entit
...
+        }
+
+        return new \Twig_Source($templateSource, $templateName);
```
Comment (kunicmarko20) [inline src/Loader/DatabaseTwigLoader.php]: Use Twig 2.x namespace?

File: src/Loader/DatabaseTwigLoader.php [hunk 17]
```diff
@@ -18,50 +18,75 @@ public function __construct(EntityManagerInterface $entityManager, string $entit
...
     {
-        $value = null;
-
         try {
-            $value = $this->entityManager
+            return $this->entityManager
                 ->getRepository($this->entity)
                 ->createQueryBuilder('t')
-                ->select('t.'.$column)
+                ->select(sprintf('t.%s', $column))
```
Comment (kunicmarko20) [inline src/Loader/DatabaseTwigLoader.php]: I understand if the character that has to be replaced is somewhere in the middle, but why if it is in the end?

File: src/Loader/DatabaseTwigLoader.php [hunk 18]
```diff
@@ -18,50 +18,75 @@ public function __construct(EntityManagerInterface $entityManager, string $entit
...
                 ->where('t.name = :name')
                 ->setMaxResults(1)
                 ->getQuery()
                 ->setParameter('name', $templateName)
                 ->getSingleScalarResult();
         } catch (NoResultException $e) {
+            return null;
```
Comment (kunicmarko20) [inline src/Loader/DatabaseTwigLoader.php]: the name is `getValue` should it error out if there is no value?

File: tests/Functional/TwigControllerTest.php [hunk 19]
```diff
@@ -27,14 +27,18 @@ class TwigControllerTest extends TestCase
...
         $this->kernel->boot();
 
-        $this->entityManager = $this->kernel->getContainer()->get('doctrine.orm.entity_manager');
+        $entityManager = $this->kernel->getContainer()->get('doctrine.orm.entity_manager');
+        assert($entityManager instanceof EntityManagerInterface);
+        $this->entityManager = $entityManager;
 
-        $this->twig = $this->kernel->getContainer()->get('twig');
+        $twig = $this->kernel->getContainer()->get('twig');
+        assert($twig instanceof \Twig_Environment);
```
Comment (kunicmarko20) [inline tests/Functional/TwigControllerTest.php]: twig namespace

File: src/Loader/DatabaseTwigLoader.php (hunk 15 above)
Comment (martin-georgiev) [reply src/Loader/DatabaseTwigLoader.php]: `Twig\Error\LoaderError` got introduced in v2.4, the goal here (for now) is to support Twig versions 2.0-2.6.

Comment (kunicmarko20) [reply src/Loader/DatabaseTwigLoader.php]: Ah, wasn't aware of that, then just ignore the namespace stuff

File: src/Loader/DatabaseTwigLoader.php (hunk 17 above)
Comment (martin-georgiev) [reply src/Loader/DatabaseTwigLoader.php]: I believe it reads better and improves consistency.
Admittedly it's an opinioned change.

File: src/DependencyInjection/Compiler/TwigLoaderPass.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (mdjward) [reply src/DependencyInjection/Compiler/TwigLoaderPass.php]: @ben-challis IMO, the entire compiler pass can be removed.  An existing compiler pass in TwigBundle which reads the `twig.loader` tags achieves the same effect.

File: src/Loader/DatabaseTwigLoader.php (hunk 5 above)
Comment (mdjward) [reply src/Loader/DatabaseTwigLoader.php]: Plus, the old names are `class_alias`'d to the new ones for BC.

File: src/DependencyInjection/Compiler/TwigLoaderPass.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (mdjward) [reply src/DependencyInjection/Compiler/TwigLoaderPass.php]: Done.

File: src/Loader/DatabaseTwigLoader.php (hunk 5 above)
Comment (mdjward) [reply src/Loader/DatabaseTwigLoader.php]: Done.

File: src/DependencyInjection/Compiler/TwigLoaderPass.php (hunk 9 above)
Comment (mdjward) [reply src/DependencyInjection/Compiler/TwigLoaderPass.php]: @ben-challis see previous comments; got rid of this entirely.

File: src/AlphaTwigBundle.php [hunk 20]
```diff
@@ -22,22 +22,22 @@ public function build(ContainerBuilder $container): void
...
 
-    private function deriveMappingDir(ContainerBuilder $container): string
+    private function resolveMappingDir(ContainerBuilder $container): string
     {
         if ($container->hasParameter(self::PARAM_TEMPLATE_MAPPING_DIR)) {
             return $container->getParameter(self::PARAM_TEMPLATE_MAPPING_DIR);
         }
 
-        return realpath(__DIR__.'/Resources/config/doctrine');
+        return (string) realpath(__DIR__.'/Resources/config/doctrine');
```
Comment (martin-georgiev) [inline src/AlphaTwigBundle.php]: @mdjward Would be better if this throws when `realpath`'s result is `false`.

File: composer.json [hunk 21]
```diff
@@ -32,45 +32,54 @@
...
-    "php": "^7.0",
+    "php": "^7.1",
     "ext-PDO": "*",
-    "doctrine/doctrine-bundle": "*",
+    "doctrine/doctrine-bundle": "^1.3.0",
     "doctrine/migrations": "*",
     "doctrine/orm": "*",
     "sonata-project/admin-bundle": "^3.0",
     "symfony/framework-bundle": "^2.7|^3.0|^4.0",
     "symfony/symfony": "^2.7|^3.0|^4.0",
```
Comment (martin-georgiev) [inline composer.json]:
```suggestion
    "symfony/symfony": "^3.4.35|^4.3|^5.0",
```

File: composer.json [hunk 22]
```diff
@@ -32,45 +32,54 @@
...
   "require": {
-    "php": "^7.0",
+    "php": "^7.1",
     "ext-PDO": "*",
-    "doctrine/doctrine-bundle": "*",
+    "doctrine/doctrine-bundle": "^1.3.0",
     "doctrine/migrations": "*",
     "doctrine/orm": "*",
     "sonata-project/admin-bundle": "^3.0",
     "symfony/framework-bundle": "^2.7|^3.0|^4.0",
```
Comment (martin-georgiev) [inline composer.json]:
```suggestion
    "symfony/framework-bundle": "^3.4.26|^4.3|^5.0",
```

File: composer.json [hunk 23]
```diff
@@ -32,45 +32,54 @@
...
     "symfony/symfony": "^2.7|^3.0|^4.0",
     "symfony/twig-bundle": "^2.7|^3.0|^4.0",
```
Comment (martin-georgiev) [inline composer.json]:
```suggestion
    "symfony/twig-bundle": "^3.4|^4.3|^5.0",
```

File: README.md (hunk 14 above)
Comment (alexlongshaw) [reply README.md]: @martin-georgiev can this be addressed please

Comment (martin-georgiev) [reply README.md]: > why this change?

merely a personal preference in the style

Comment (alexlongshaw) [reply README.md]: Please can you revert to how it was before. Can't see any other example of the 3 character =

Comment (martin-georgiev) [inline README.md]:
```suggestion
==========
```

File: README.md [hunk 24]
```diff
@@ -1,12 +1,16 @@
...
+[![Total Downloads](https://poser.pugx.org/lendable/twig-bundle/downloads)](https://packagist.org/packages/lendable/twig-bundle)
 
 TwigBundle
 ==========
 
 Symfony bundle that allows Twig templates to be loaded from database store.
 
 Install
-=======
+===
```
Comment (martin-georgiev) [inline README.md]:
```suggestion
=======
```

File: README.md [hunk 25]
```diff
@@ -42,5 +46,5 @@ parameters:
 ```
 
 License
-=======
+===
```
Comment (martin-georgiev) [inline README.md]:
```suggestion
=======
```
